azure/services/managedclusters/managedclusters.go

```diff
@@ -147,3 +157,75 @@ func (s *Service) Delete(ctx context.Context) error {
 func (s *Service) IsManaged(ctx context.Context) (bool, error) {
 	return true, nil
 }
+
+// ReconcileKubeconfig will reconcile admin kubeconfig and user kubeconfig.
+/*
+  Returns the admin kubeconfig and user kubeconfig
+  If aad is enabled a user kubeconfig will also get generated and stored in the secret <cluster-name>-kubeconfig-user
+  If we disable local accounts for aad clusters we do not have access to admin kubeconfig, hence we need to create
+  the admin kubeconfig by authenticating with the user credentials and retrieving the token for kubeconfig.
+  The token is used to create the admin kubeconfig.
+  The user needs to ensure to provide service principle with admin aad privileges.
+*/
+func (s *Service) ReconcileKubeconfig(ctx context.Context, managedClusterSpec azure.ResourceSpecGetter) (userKubeConfigData []byte, adminKubeConfigData []byte, err error) {
+	if s.Scope.IsAADEnabled() {
+		if userKubeConfigData, err = s.GetUserKubeconfigData(ctx, managedClusterSpec); err != nil {
+			return nil, nil, errors.Wrap(err, "error while trying to get user kubeconfig")
+		}
+	}
+
+	if s.Scope.IsLocalAcountsDisabled() {
+		userKubeconfigWithToken, err := s.GetUserKubeConfigWithToken(userKubeConfigData, ctx, managedClusterSpec)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "error while trying to get user kubeconfig with token")
+		}
+		return userKubeconfigWithToken, userKubeConfigData, nil
+	}
+
+	adminKubeConfigData, err = s.GetCredentials(ctx, managedClusterSpec.ResourceGroupName(), managedClusterSpec.ResourceName())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to get credentials for managed cluster")
+	}
+	return adminKubeConfigData, userKubeConfigData, nil
+}
+
+// GetUserKubeconfigData gets user kubeconfig when aad is enabled for the aad clusters.
+func (s *Service) GetUserKubeconfigData(ctx context.Context, managedClusterSpec azure.ResourceSpecGetter) ([]byte, error) {
+	kubeConfigData, err := s.GetUserCredentials(ctx, managedClusterSpec.ResourceGroupName(), managedClusterSpec.ResourceName())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get credentials for managed cluster")
+	}
+	return kubeConfigData, nil
+}
+
+// GetUserKubeConfigWithToken returns the kubeconfig with user token, for capz to create the target cluster.
+func (s *Service) GetUserKubeConfigWithToken(userKubeConfigData []byte, ctx context.Context, managedClusterSpec azure.ResourceSpecGetter) ([]byte, error) {
+	tokenClient, err := token.NewClient(s.Scope)
+	if err != nil {
+		return nil, errors.Wrap(err, "error while getting aad token client")
+	}
+
+	token, err := tokenClient.GetAzureActiveDirectoryToken(ctx, aadResourceID)
```

Comment:
I don't see any major blockers with ASO. It might be clunky to add this token to the user kubeconfig if ASO doesn't do that for us, but I don't think that would be any more involved than this already is.

@LochanRn Would you be able to help test this feature in #4069 when I resolve conflicts there?